### DISCUSSION TITLE
Management UI for custom directories (custom central servers)

### DISCUSSION THREAD

**Discussion by jp8** [jamulussoftware/jamulus#1899] [UI Related]
I am opening this issue as requested in code review

https://github.com/jamulussoftware/jamulus/pull/1894#discussion_r660755323

of pull request #1894 .

Something needs to be done about the UI for management of custom central servers.  The initial suggestion from @pljones is to update the list on the connection dialog, to show the actual directory server selected or changed on the settings dialog  (i.e. remember where we were and see if we can find it in the new list).

It is likely related to discussion #1774

**Comment by gilgongo** [comment]
There were two pull requests https://github.com/jamulussoftware/jamulus/pull/1884 and #1894, and I thought we had agreed on the approach taken in the latter. But I see this https://github.com/jamulussoftware/jamulus/pull/1894#issuecomment-874341326 so I've lost track of the status of all this now is it merged?

**Reply by jp8** [reply]
The UI for selecting a custom directory on the Connect dialog (#1894) got merged.
The UI for adding / deleting / re-ordering custom directories is up for discussion.  There were a lot of good but sometimes conflicting ideas of how the UI for management of might be improved.

**Comment by pljones** [comment]
With the latest change, the situation is:
- Connect lists all known directory servers.  This means the predefined set (_somegenre_.jamulus.io) plus all known custom entries
- Custom entries can currently be added by entering them in the Settings->Advanced tab

The aim here is to agree some improvements to the latter mechanism (i.e. how to manage the custom list).  Whether it remains where it is now or not is open to discussion.

My opinion, as a starter, is that the current location is fine.  Maintaining the list doesn't happen often, compared with connecting to a server for a jam session.  It makes sense under Settings.  (Whether "Advanced" is the right place or not?  Well... let's see.)  It could do with the abilities to:
- Add an entry
- Delete an entry
- Reorder entries in the list
- View the server list for an entry

Ideally none of the above should affect what the Connect dialog is displaying, with the exception of deleting a custom entry where the Connect dialog was displaying _that_ entry.

**Reply by gilgongo** [reply]
I see. Adding and deleting entries is not an issue in the current implementation, I assume? Can you explain what you mean by "View the server list for an entry"? Also why anyone would need to re-order the list? How many custom directories are we talking about here? 

The reason I ask is that in general I think we should stick to not confusing or inconveniencing the majority of users for the sake of a minority. The proportion of Jamulus users who need to "manage" directories with any kind of frequency is extremely tiny (although we are talking about an "advanced" tab, so maybe it's not too bad).

**Reply by dcorson-ticino-com** [reply]
I think the combobox on the advanced settings tab already does all that is needed.
You can add and delete servers.
If reordering is deemed important it should be in the connect window, in settings there is no need.
I don't know why we would need to view the server list of an entry in the settings, the connect window is where that happens.

**Reply by jp8** [reply]
> Ideally none of the above should affect what the Connect dialog is displaying, with the exception of deleting a custom entry where the Connect dialog was displaying _that_ entry.

I need to point out that as a result of pull request #1894 the connect dialog now switches back to the default directory, as soon as someone begins modifying (on the settings dialog) the list of custom servers.

One of the reasons for this discussion is to decide if and how that should be addressed.

**Reply by gilgongo** [reply]
> One of the reasons for this discussion is to decide if and how that should be addressed.

The desired behaviour would be to keep the list on the connection dialogue unchanged, I assume. And if you delete the one that's defaulting, the next one takes its place?

**Reply by jp8** [reply]
Yes that was the view of @pljones but I requested the task to be separated out because there is some discussion around exactly what functionality should be offered for custom directory management, and how it should look in the UI and I wanted that to settle a bit before thinking about the best way for the settings dialog to inform the connection dialog that the list of custom directories has been re-ordered.

**Reply by jp8** [reply]
> The desired behaviour would be to keep the list on the connection dialogue unchanged, I assume. And if you delete the one that's defaulting, the next one takes its place?

I've now done an implementation that keeps the current selection (whatever that is) on the connection dialog unless that selection is a custom directory that gets deleted, in which case, the connection dialog goes to AT_DEFAULT.

I hope that is ok ?

Shall I squash it to a single commit and then make a new pull request ?

**Comment by jp8** [comment]
I have created pull request #1914, to keep the existing combo box selection unless it has been deleted.
